src/Doc/Man-Pages/kmetis.1
.TH KMETIS 1
.SH NAME
kmetis \- multilevel k-way graph partitioner
.SH SYNOPSIS
.B kmetis
.I GraphFile
.I Nparts
.SH DESCRIPTION
.B kmetis
partitions the graph stored in
.I GraphFile
into 
.I Nparts 
using an algorithm that is based on multilevel k-way partition.
The result of the partition is stored in a file named
.I GraphFile.part.Nparts.

.PP
.I Nparts
can be any number greater than one and less than the number of vertices in the
graph. 


.SH FORMAT OF INPUT FILE
A graph 
.B G = (V, E) 
with 
.B n 
vertices and 
.B m 
edges is stored in a plain
text file that contains 
.B n+1 
lines.
The first line contains information about the size and the type of the graph,
while the remaining 
.B n 
lines contain information for each vertex of 
.B G
.
Any line that starts with 
.B % 
is a comment line and is skipped.

.PP
The first line contains either two or three integers. The first integer
is the number of vertices, the second is the number of edges, and
the third integer 
.B fmt 
contains information about the
type of the graph. In particular, depending on the value of 
.B fmt, 
the graph can have weights on the edges, vertices, or both.
In the case that the graph is unweighted, (ie., all vertices and edges have the
same weight),
.B fmt
is omitted.

.PP
After this first line, the remaining 
.B n 
lines store the adjacency lists for each
vertex-one line per vertex.
In particular, the 
.B i^th 
line (excluding comment lines) contains the adjacency list of the 
.B (i-1)^st 
vertex.
Note that the numbering starts from 1 (not from 0 as it is done in C).

.PP
The simplest format for a graph is when the weight of all vertices and the weight
of all the edges is the same.
However, there are cases in which the edges in 
.B G
have different weights. 
In that case, the adjacency list of each vertex contains in addition to the vertices
that is connected with, the weight of the edges.
If 
.B v
has 
.B k 
vertices adjacent to it, then the graph file contains
.B 2*k
numbers, each pair of numbers stores the vertex that 
.B v 
is connected to, and the
weight of the edge. Note, edge weights are integer quantities.
Furthermore, note that the 
.B fmt 
parameter is equal to 
.B 1
indicating the fact that the graph has weights on the edges.

.PP
Finally, in addition to having weights on the edges, weights on the vertices
are also allowed.
In this case, the value of 
.B fmt
is equal to
.B 11
and each line of the graph file first stores the weight of the vertex, 
and then the weighted adjacency list.
As was the case with edge weights, vertex weights are integer quantities.


.SH FORMAT OF OUTPUT FILE
The partition file of a graph with 
.B n
vertices, consists of 
.B n
lines with a single number per line. 
The 
.B i^th 
line of the file contains the partition number that the 
.B i^th 
vertex belongs to. Partition numbers start from 0, and partitions are assigned 
in a hypercube fashion.

.SH ADDITIONAL DOCUMENTATION
A postscript copy of the manual of 
.B METIS
that describes 
.B kmetis
and other similar algorithms is available on WWW at:
.B http://www.cs.umn.edu/~karypis/metis/manual.ps


.SH AUTHOR
George Karypis
.SH DIAGNOSTICS
Diagnostic messages are intended to be self-explanatory.
.SH "SEE ALSO"
metis(1), pmetis(1), KMETIS(3)
.SH BUGS
If you find any bugs please report them to
.B karypis@cs.umn.edu
.


src/CHANGES

METIS 3.0.6, 1/28/98
-------------------------------------------------------------------------------
  - Fixed some problems when too many partitions were asked, and each partition
    end up having 0 vertices
  - Fixed some bugs in the I/O routines
  - Added support for the g77 compiler under Linux


METIS 3.0.5, 12/22/97
-------------------------------------------------------------------------------
  - Fixed problems on 64-bit architectures (eg., -64 option on SGIs).
  - Added some options in Makefile.in


METIS 3.0.4, 12/1/97
-------------------------------------------------------------------------------
  Fixed a memory leak in the ordering code.


METIS 3.0.3, 11/5/97
-------------------------------------------------------------------------------
  This is mostly a bug-fix release with just a few additions

  Added functionality
    - Added support for quadrilateral elements.
    - Added a routine METIS_EstimateMemory that estimates the amount of
      memory that will be allocated by METIS. This is useful in determining
      if a problem can run on your system.
    - Added hooks to allow PARMETIS to use the orderings produced by METIS.
      This is hidden from the user but it will be used in the next release
      of PARMETIS.

  Bug-fixes
    - Fixed a bug related to memory allocation. This should somewhat reduce the 
      overall memory used by METIS.
    - Fixed some bugs in the 'graphchk' program in the case of weighted graphs.
    - Removed some code corresponding to unused options.
    - Fixed some minor bugs in the node-refinement code
      


-------------------------------------------------------------------------------
METIS 3.0 contains a number of changes over METIS 2.0.
The major changes are the following:

  General Changes
  ---------------
    1.  Added code to directly partition finite element meshes.

    2.  Added code to convert finite element meshes into graphs so they
        can be used by METIS.

    1.  The names, calling sequences, and options  of the routines in 
        METISlib have been changed.

    2.  Better support has been added for Fortran programs.

    3.  Eliminated the 'metis' program. The only way to tune METIS's
        behavior is to use METISlib.

    4.  Improved memory management. METIS should now only abort if truly
        there is no more memory left in the system.


  Graph Partitioning
  ------------------
    1.  Added partitioning routines that can be used to compute a partition 
        with prescribed partition weights. For example, they can be used to 
        compute a 3-way partition such that partition 1 has 50% of the weight, 
        partition 2 has 20% of the way, and partition 3 has 30% of the weight. 

    2.  Improved the speed of the k-way partitioning algorithm (kmetis). The
        new code has better cache locality which dramatically improves the 
        speed for large graphs. A factor of 4 speedup can be obtained for
        certain graphs. METIS can now partition a 4 million node graph
        in well under a minute on a MIPS R10000.

    3.  Eliminated some of the options that were seldom used. 
  

  Fill-Reducing Orderings
  ----------------------
    1.  Added a node based ordering code `onmetis' that greatly improves 
        ordering quality.

    2.  Improved the quality of the orderings produced by the original
        edge-based ordering code (it is now called 'oemetis').

    3.  METIS can now analyze the graph and try to compress together 
        nodes with identical sparsity pattern. For some problems, this 
        significantly reduces ordering time 

    4.  METIS can now prune dense columns prior to ordering. This can be
        helpful for LP matrices.
        

  Mesh Partitioning
  -----------------
    1.  METIS can now directly partition the element node array of finite
        element meshes. It produces two partitioning vectors. One for the
        elements and one for the nodes. METIS supports the following 
        elements: triangles, tetrahedra, hexahedra


  Mesh-To-Graph Conversion Routines
  ---------------------------------
    1.  METIS now includes a number of mesh conversion functions that can 
        be used to create the dual and nodal graphs directly from the 
        element connectivity arrays. These are highly optimized routines. 





src/Lib/kmetis.c
/*
 * Copyright 1997, Regents of the University of Minnesota
 *
 * kmetis.c
 *
 * This file contains the top level routines for the multilevel k-way partitioning
 * algorithm KMETIS.
 *
 * Started 7/28/97
 * George
 *
 * $Id: kmetis.c,v 1.2 1998/02/10 19:34:22 karypis Exp $
 *
 */

#include <metis.h>


/*************************************************************************
* This function is the entry point for KMETIS
**************************************************************************/
void METIS_PartGraphKway(int *nvtxs, idxtype *xadj, idxtype *adjncy, idxtype *vwgt, 
                         idxtype *adjwgt, int *wgtflag, int *numflag, int *nparts, 
                         int *options, int *edgecut, idxtype *part)
{
  int i;
  float *tpwgts;

  tpwgts = fmalloc(*nparts, "KMETIS: tpwgts");
  for (i=0; i<*nparts; i++) 
    tpwgts[i] = 1.0/(1.0*(*nparts));

  METIS_WPartGraphKway(nvtxs, xadj, adjncy, vwgt, adjwgt, wgtflag, numflag, nparts, 
                       tpwgts, options, edgecut, part);

  free(tpwgts);
}


/*************************************************************************
* This function is the entry point for KWMETIS
**************************************************************************/
void METIS_WPartGraphKway(int *nvtxs, idxtype *xadj, idxtype *adjncy, idxtype *vwgt, 
                          idxtype *adjwgt, int *wgtflag, int *numflag, int *nparts, 
                          float *tpwgts, int *options, int *edgecut, idxtype *part)
{
  int i, j;
  GraphType graph;
  CtrlType ctrl;

  if (*numflag == 1)
    Change2CNumbering(*nvtxs, xadj, adjncy);

  SetUpGraph(&graph, OP_KMETIS, *nvtxs, xadj, adjncy, vwgt, adjwgt, *wgtflag);

  if (options[0] == 0) {  /* Use the default parameters */
    ctrl.CType = KMETIS_CTYPE;
    ctrl.IType = KMETIS_ITYPE;
    ctrl.RType = KMETIS_RTYPE;
    ctrl.dbglvl = KMETIS_DBGLVL;
  }
  else {
    ctrl.CType = options[OPTION_CTYPE];
    ctrl.IType = options[OPTION_ITYPE];
    ctrl.RType = options[OPTION_RTYPE];
    ctrl.dbglvl = options[OPTION_DBGLVL];
  }
  ctrl.optype = OP_KMETIS;
  ctrl.CoarsenTo = 20*(*nparts);
  ctrl.maxvwgt = 1.5*(idxsum(*nvtxs, graph.vwgt)/ctrl.CoarsenTo);

  InitRandom();

  AllocateWorkSpace(&ctrl, &graph, *nparts);

  IFSET(ctrl.dbglvl, DBG_TIME, InitTimers(&ctrl));
  IFSET(ctrl.dbglvl, DBG_TIME, starttimer(ctrl.TotalTmr));

  *edgecut = MlevelKWayPartitioning(&ctrl, &graph, *nparts, part, tpwgts, 1.03);

  IFSET(ctrl.dbglvl, DBG_TIME, stoptimer(ctrl.TotalTmr));
  IFSET(ctrl.dbglvl, DBG_TIME, PrintTimers(&ctrl));

  FreeWorkSpace(&ctrl, &graph);

  if (*numflag == 1)
    Change2FNumbering(*nvtxs, xadj, adjncy, part);
}


/*************************************************************************
* This function takes a graph and produces a bisection of it
**************************************************************************/
int MlevelKWayPartitioning(CtrlType *ctrl, GraphType *graph, int nparts, idxtype *part, float *tpwgts, float ubfactor)
{
  int i, j, nvtxs, tvwgt, tpwgts2[2];
  GraphType *cgraph;
  int wgtflag=3, numflag=0, options[10], edgecut;

  cgraph = Coarsen2Way(ctrl, graph);

  IFSET(ctrl->dbglvl, DBG_TIME, starttimer(ctrl->InitPartTmr));
  AllocateKWayPartitionMemory(ctrl, cgraph, nparts);

  options[0] = 1; 
  options[OPTION_CTYPE] = MATCH_SHEMKWAY;
  options[OPTION_ITYPE] = IPART_GGPKL;
  options[OPTION_RTYPE] = RTYPE_FM;
  options[OPTION_DBGLVL] = 0;

  METIS_WPartGraphRecursive(&cgraph->nvtxs, cgraph->xadj, cgraph->adjncy, cgraph->vwgt, 
                            cgraph->adjwgt, &wgtflag, &numflag, &nparts, tpwgts, options, 
                            &edgecut, cgraph->where);

  IFSET(ctrl->dbglvl, DBG_TIME, stoptimer(ctrl->InitPartTmr));
  IFSET(ctrl->dbglvl, DBG_IPART, printf("Initial %d-way partitioning cut: %d\n", nparts, edgecut));

  IFSET(ctrl->dbglvl, DBG_KWAYPINFO, ComputePartitionInfo(cgraph, nparts, cgraph->where));

  RefineKWay(ctrl, graph, cgraph, nparts, tpwgts, ubfactor);

  idxcopy(graph->nvtxs, graph->where, part);

  GKfree(&graph->gdata, &graph->rdata, -1);

  return graph->mincut;

}

